apps/browser/qml/pages/components/TabView.qml

```diff
@@ -40,131 +32,144 @@ SilicaGridView {
-    currentIndex: -1
-    header: PageHeader {
-        //: Tabs
-        //% "Tabs"
-        title: qsTrId("sailfish_browser-he-tabs")
-    }
-    footer: spacer
-    cellHeight: thumbnailHeight + Theme.paddingLarge
-    cellWidth: thumbnailWidth + Theme.paddingLarge
-
-    delegate: TabItem {
-        id: tabItem
-
-        enabled: !closingAllTabs
-        opacity: enabled ? 1.0 : 0.0
-        Behavior on opacity { FadeAnimator {}}
-
-        width: thumbnailWidth
-        height: thumbnailHeight
-
-        GridView.onAdd: AddAnimation {
-            target: tabItem
-        }
-        GridView.onRemove: RemoveAnimation {
-            target: tabItem
-        }
-    }
-
-    // Behind tab delegates
-    children: [
-        PrivateModeTexture {
-            z: -1
-            visible: opacity > 0.0
-            opacity: privateMode ? 1.0 : 0.0
-
-            Behavior on opacity { FadeAnimation {} }
-        },
+    anchors.fill: parent
 
-        MouseArea {
+    header: Private.TabBar {
+        id: header
+        model: headerModel
+        Shared.Background {
+            anchors.fill: parent
             z: -1
-            width: tabView.width
-            height: tabView.height
-            onClicked: hide()
         }
-    ]
-
-    ConfigurationValue {
-        id: showCloseAllAction
-        key: "/apps/sailfish-browser/settings/show_close_all"
-        defaultValue: true
     }
 
-    PullDownMenu {
-        id: pullDownMenu
+    model: [ persistentTabView, privateTabView ]
 
-        flickable: tabView
+    onCurrentIndexChanged: {
+        // TODO: Add remorsePopup
+        // if (remorsePopup) {
+        //     remorsePopup.trigger()
+        // }
+        privateMode = currentIndex !== 0
+    }
+    Component.onCompleted: {
+        persistentIcon.updateGrubImage()
+        privateIcon.updateGrubImage()
+        currentIndex = privateMode
+    }
 
-        MenuItem {
-            text: tabView.privateMode ?
-                    //: Menu item switching back to normal browser
-                    //% "Normal browsing"
-                    qsTrId("sailfish_browser-me-normal_browsing") :
-                    //: Menu item switching to private browser
-                    //% "Private browsing"
-                    qsTrId("sailfish_browser-me-private_browsing")
-            onDelayedClick: {
-                if (remorsePopup) {
-                    remorsePopup.trigger()
-                }
+    PrivateModeTexture {
+        z: -1
+        visible: opacity > 0.0
+        opacity: privateMode ? 1.0 : 0.0
 
-                tabView.privateMode = !tabView.privateMode
+        Behavior on opacity { FadeAnimation {} }
+    }
 
+    Component {
+        id: persistentTabView
+
+        Private.TabItem {
+            TabGridView {
+                id: _persistentTabView
+                privateMode: false
+                portrait: tabs.portrait
+                model: !tabs.privateMode ? tabs.tabModel : null
```

Comment:
Using the same `tabModel` and switching between them like this seems to be causing problems for the animation when swiping between the normal and private tab listings. The items appear suddenly half way through the swipe.

To get this to work nicely, I think it'll be necessary to maintain both models simultaneously and separately here.

Reply:
I agree, corrected